packages/flutter/lib/src/material/theme_data.dart

```diff
@@ -1635,3 +1674,141 @@ class _FifoCache<K, V> {
     return _cache[key] = loader();
   }
 }
+
+/// Defines the visual density of user interface components.
+///
+/// Density, in the context of a UI, is the vertical and horizontal
+/// "compactness" of the components in the UI. It is unitless, since it means
+/// different things to different UI components.
+///
+/// For example, for buttons, it affects the spacing around the child of the
+/// button. For lists, it affects the distance between baselines of entries in
+/// the list. For chips, it only affects the vertical size, not the horizontal
+/// size.
+class VisualDensity extends Diagnosticable {
+  /// A const constructor for [VisualDensity].
+  ///
+  /// All of the arguments must be non-null, and [horizontal] and [vertical]
+  /// must be in the interval between [minimumDensity] and [maximumDensity],
+  /// inclusive.
+  const VisualDensity({
+    this.horizontal = 0.0,
+    this.vertical = 0.0,
+  }) : assert(horizontal != null),
+       assert(vertical != null),
+       assert(vertical <= maximumDensity),
+       assert(vertical >= minimumDensity),
+       assert(horizontal <= maximumDensity),
+       assert(horizontal >= minimumDensity);
+
+  /// The minimum allowed density.
+  static const double minimumDensity = -4.0;
+
+  /// The maximum allowed density.
+  static const double maximumDensity = 4.0;
+
+  /// The default profile for [VisualDensity] in [ThemeData].
+  ///
+  /// This default value represents a visual density that is less dense than
+  /// either [comfortable] or [compact].
+  static const VisualDensity standard = VisualDensity();
+
+  /// The profile for a "comfortable" interpretation of [VisualDensity].
+  ///
+  /// Individual components will interpret the density value independently,
+  /// making themselves more visually dense than [standard] and less dense than
+  /// [compact] to different degrees based on the Material Design specification
+  /// of the "comfortable" setting for their particular use case.
+  static const VisualDensity comfortable = VisualDensity(horizontal: -1.0, vertical: -1.0);
+
+  /// The profile for a "compact" interpretation of [VisualDensity].
+  ///
+  /// Individual components will interpret the density value independently,
+  /// making themselves more visually dense than [standard] and [comfortable] to
+  /// different degrees based on the Material Design specification of the
+  /// "comfortable" setting for their particular use case.
+  static const VisualDensity compact = VisualDensity(horizontal: -2.0, vertical: -2.0);
+
+  /// Copy the current [VisualDensity] with the given values replacing the
+  /// current values.
+  VisualDensity copyWith({
+    double horizontal,
+    double vertical,
+  }) {
+    return VisualDensity(
+      horizontal: horizontal ?? this.horizontal,
+      vertical: vertical ?? this.vertical,
+    );
+  }
+
+  /// The horizontal visual density of UI components.
```

Comment:
Should we say that the default is 0 is the valid range is -4 to 4? Likewise for horizontal

Reply:
Done.